src/libawkward/typedbuilder/UnknownFormBuilder.cpp

```diff
@@ -0,0 +1,65 @@
+// BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+#define FILENAME(line) FILENAME_FOR_EXCEPTIONS("src/libawkward/builder/TypedArrayBuilder.cpp", line)
+
+#include "awkward/typedbuilder/UnknownFormBuilder.h"
+#include "awkward/typedbuilder/TypedArrayBuilder.h"
+
+namespace awkward {
+
+  ///
+  UnknownFormBuilder::UnknownFormBuilder(const FormPtr& form)
+    : form_(form),
+      vm_empty_command_("( This does nothing. )\n"),
+      vm_error_("s\"Unknown Form Builder error\"") { }
```

Comment:
The "unknown" type can only come from an EmptyArray/EmptyForm, right? If there is such a thing in the user-provided form, then any attempt to insert data at this level should be an error. For instance, if the form is `ListOffsetForm(ListOffsetForm(EmptyForm()))`, then it would be possible to make

```
[[], [[], [], []], [[], []], [], [[]]]
```

with an appropriate sequence of `begin_list`/`end_list`, but not

```
[[], [[], [], []], [[], []], [], [[1]]]
```

This is different from the use of "unknown" in ArrayBuilder. Every ArrayBuilder starts as unknown and any operation is valid on it because ArrayBuilder changes its type; TypedArrayBuilder doesn't change type, so no operation is valid on an unknown type.

(It wouldn't be particularly useful to use EmptyForms with TypedArrayBuilder; it's just here for completeness. The whole concept of unknown type, EmptyArray, and EmptyForm was added only for ArrayBuilder to have a starting point, and since it shouldn't be an error for some lists to never be filled, unknown type is not just an internal detail of ArrayBuilder.)

Reply:
I've added this test and removed "unknown" array builder.